api/src/test/java/io/grpc/ContextTest.java

```diff
@@ -935,6 +949,104 @@ public Context current() {
     }
   }
 
+  @Test
+  public void customStorage() throws Exception {
+    JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+
+    Map<String, ByteArrayOutputStream> classBytes = new HashMap<>();
+    JavaFileManager fileManager = new ForwardingJavaFileManager<StandardJavaFileManager>(
+        compiler.getStandardFileManager(null, null, null)) {
+      @Override
+      public JavaFileObject getJavaFileForOutput(Location location, String className,
+          JavaFileObject.Kind kind, FileObject sibling) {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        classBytes.put(className, outputStream);
+        return new SimpleJavaFileObject(
+            URI.create("string:///" + className.replace('.', '/') + kind.extension), kind) {
+          @Override
+          public OutputStream openOutputStream() {
+            return outputStream;
+          }
+        };
+      }
+    };
+
+    JavaFileObject sourceFile = new SimpleJavaFileObject(URI.create(
+        "string:///ContextStorageOverride.java"),
+        JavaFileObject.Kind.SOURCE) {
+      @Override
+      public CharSequence getCharContent(boolean ignoreEncodingErrors) {
+        return "package io.grpc.override;\n"
+            + "\n"
+            + "import io.grpc.Context;\n"
+            + "import io.grpc.Context.Storage;\n"
+            + "\n"
+            + "public class ContextStorageOverride extends Storage {\n"
+            + "  public static Context toAttach = null;\n"
+            + "  public static Context toDetach = null;\n"
+            + "  public static Context toRestore = null;\n"
+            + "  public static Context current = null;\n"
+            + "\n"
+            + "  @Override\n"
+            + "  public Context doAttach(Context toAttach) {\n"
+            + "    ContextStorageOverride.toAttach = toAttach;\n"
+            + "    return current();\n"
+            + "  }\n"
+            + "\n"
+            + "  @Override\n"
+            + "  public void detach(Context toDetach, Context toRestore) {\n"
+            + "    ContextStorageOverride.toDetach = toDetach;\n"
+            + "    ContextStorageOverride.toRestore = toRestore;\n"
+            + "  }\n"
+            + "\n"
+            + "  @Override\n"
+            + "  public Context current() {\n"
+            + "    return current;\n"
+            + "  }\n"
+            + "}\n";
+      }
+    };
+
+    boolean result = compiler.getTask(null, fileManager, null, null, null,
+            Collections.singletonList(sourceFile)).call();
+    if (!result) {
+      fail("Compilation failed");
+      return;
+    }
+
+    ClassReplacingClassLoader classLoader = new
+        ClassReplacingClassLoader(
+        getClass().getClassLoader(), Pattern.compile("io\\.grpc\\.[^.]+"), classBytes);
+
+    Class<?> storageClass = classLoader.loadClass("io.grpc.override.ContextStorageOverride");
```

Comment:
Right. The approach I was pointing out would use regular Java code to look at the inner field, but would load that code within the faked classloader as well. It then has direct access. You then use an API like Runnable/Callable that is shared between the class loaders to communicate with the test code.

(Not that any of this matters, since I think we're at a dead-end since avoiding javac is too much of a pain but we also don't want the javac dependency.)